lib/model/model.go

```diff
@@ -1068,81 +1195,81 @@ func (m *model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
-					// information we have from them before accepting their
-					// index, which will presumably be a full index.
-					fs.Drop(deviceID)
-				} else if dev.IndexID != theirIndexID {
-					// The index ID we have on file is not what they're
-					// announcing. They must have reset their database and
-					// will probably send us a full index. We drop any
-					// information we have and remember this new index ID
-					// instead.
-					l.Infof("Device %v folder %s has a new index ID (%v)", deviceID, folder.Description(), dev.IndexID)
-					fs.Drop(deviceID)
-					fs.SetIndexID(deviceID, dev.IndexID)
-				} else {
-					// They're sending a recognized index ID and will most
-					// likely use delta indexes. We might already have files
-					// that we need to pull so let the folder runner know
-					// that it should recheck the index data.
-					m.fmut.RLock()
-					if runner := m.folderRunners[folder.ID]; runner != nil {
-						defer runner.SchedulePull()
-					}
-					m.fmut.RUnlock()
+		// This is the other side's description of what it knows
+		// about us. Lets check to see if we can start sending index
+		// updates directly or need to send the index from start...
+
+		if hasDeviceLocal && ccDeviceLocal.IndexID == myIndexID && ccDeviceLocal.MaxSequence <= mySequence {
+			// They say they've seen our index ID before and their
+			// max sequence is consistent with ours, so we can
+			// send a delta update only.
+			l.Debugf("Device %v folder %s is delta index compatible (mlv=%d)", deviceID, folder.Description(), ccDeviceLocal.MaxSequence)
+			startSequence = ccDeviceLocal.MaxSequence
+		} else if !hasDeviceLocal {
+			l.Debugf("Device %v folder %s sent no info about us", deviceID, folder.Description())
+		} else if ccDeviceLocal.IndexID == myIndexID {
+			// Safety check above failed: They claim to have more or newer
+			// index data than we have - either we have lost
+			// index data, or reset the index without resetting
+			// the IndexID, or something else weird has
+			// happened. We send a full index to reset the
+			// situation.
+			l.Infof("Device %v folder %s is delta index compatible, but seems out of sync with reality", deviceID, folder.Description())
+		} else if ccDeviceLocal.IndexID != 0 {
+			// They say they've seen an index ID from us, but it's
+			// not the right one. Either they are confused or we
+			// must have reset our database since last talking to
+			// them. We'll start with a full index transfer.
+			l.Infof("Device %v folder %s has mismatching index ID for us (%v != %v)", deviceID, folder.Description(), ccDeviceLocal.IndexID, myIndexID)
+		}
+
+		// This is the other side's description of themselves. We
+		// check to see that it matches the IndexID we have on file,
+		// otherwise we drop our old index data and expect to get a
+		// completely new set.
+
+		switch {
+		case !hasDevice:
+			l.Debugf("Device %v folder %s sent no info about themselves", deviceID, folder.Description())
+			fallthrough
```

Comment:
All this hasDevice hasDeviceLocal stuff seems just invalid state. I suspect we didn't allow it before, and now we're starting to handle this. Why?

Reply:
We did not abort before when those were missing. However it looks like we just didn't check and assumed it would be always there. So it might indeed be sensible to throw an error if any of this is missing, but that's a change from previous behaviour (could do a separate PR to do it there, the less happening in this monstrosity of a PR the better).